# Algoritmo Genetico

In [34]:
import numpy as np

Construimos la funcion objetivo, usa la funcion Rosenbrock.

In [35]:
#Rosenbrock
fob = lambda x: -1*(x[0]**4+3*x[0]**3*x[1]+\
                6*x[0]**2*x[1]**2+3*x[0]*x[1]**3+\
                x[1]**4-4*(x[0]-x[1])*(x[0]+3*x[1]))

Codigo del algoritmo genetivo visto en la diapositivas.

In [36]:
#Generar poblacion
N = 100
P = 2
X0 = 4*np.random.rand(N,P)-2 #Poblacion inicial
X = X0.copy()
#Evaluar poblacion
aptitud = []
for k in range(N):
    aptitud.append(fob(X[k,:]))

aptitud = np.array(aptitud)
indSort = np.argsort(aptitud)

print(max(aptitud), min(aptitud))

for k in range(100):
  #Ruleta
  ValoresRul = aptitud - aptitud.min() + 0.1
  S = sum(ValoresRul)
  r = np.random.rand()*S
  sumasPar = np.cumsum(ValoresRul)

  ind = np.argwhere(sumasPar>=r)[0][0]


  sinInd = list(range(N))
  sinInd.remove(ind)
  S2 = sum(ValoresRul[sinInd])
  r = np.random.rand()*S2
  sumasPar = np.cumsum(ValoresRul[sinInd])

  ind2=np.argwhere(sumasPar>=r)[0][0]


  #cruce
  alpha = np.random.rand()
  hijo = alpha*X[ind,:]+(1-alpha)*X[ind2,:]

  #Mutacion
  if np.random.rand()>0.5:
    cuantos_par = np.random.randint(X.shape[1]+1)
    indHijoMutar = np.random.permutation(X.shape[1])[:cuantos_par]
    hijo[indHijoMutar] += np.random.randn(cuantos_par)

  #Evolucion
  X[indSort[0],:] = hijo
  aptitud[indSort[0]] = fob(hijo)
  indSort = np.argsort(aptitud)

bestInd = np.argmax(aptitud)
mejorX = X[bestInd,:]
mejorAptitud = aptitud[bestInd]
print("La solucion es ",mejorX," con aptitud: ",mejorAptitud)

3.0798106918428676 -108.67808805095473
La solucion es  [-1.18792712 -0.0382197 ]  con aptitud:  3.794188884532844


Buscamos en una grilla (grid) el optimo.

In [37]:
x1 = np.linspace(-2,2,100)
x2 = np.linspace(-2,2,100)
x1v, x2v = np.meshgrid(x1,x2)

In [38]:
A = np.hstack((x1v.reshape(100*100,1),x2v.reshape(100*100,1)))

In [39]:
aptitudFB = []
for i in range(A.shape[0]):
  aptitudFB.append(fob(A[i,:]))
aptitudFB = np.asarray(aptitudFB)

In [40]:
np.max(aptitudFB),A[np.argmax(aptitudFB),:]

(np.float64(4.092306317199881), array([-1.39393939, -0.06060606]))

Segun la grilla, el maximo es 4.09 ubicado en $x_1=-1.39$, $x_2=-0.061$.

# PSO - Enjambre de Particulas

In [52]:
#Este codigo del PSO esta hecho para maximizar
Particulas = X0.copy() #misma poblacion inicial del genetico

l_inf = np.array([-2.,-2.])
l_sup = np.array([2.,2.])

# Ejemplo para minimizacion
#funobj = lambda x: np.sum(x**2) #Funcion a minimizar
#fob = lambda x: -funobj(x) #Se invierte para maximizar

F = np.apply_along_axis(fob,1,Particulas)

pbest = Particulas.copy()
Fpbest = F.copy()
indBest = np.argmax(F)
gbest = Particulas[indBest,:].copy()
Fgbest = F[indBest]
Iteraciones = 100
c1, c2 = 0.5, 0.5
Velocidad = np.zeros_like(Particulas)
for i in range(Iteraciones):
  r1 = np.random.rand(1,P)
  r2 = np.random.rand(1,P)
  #Calculo de la velocidad para todas las particulas
  Velocidad += c1*r1*(pbest-Particulas)+c2*r2*(gbest-Particulas)
  #Actualizacion posicion de todas las particulas
  Particulas += Velocidad
  Particulas = np.clip(Particulas,l_inf,l_sup)
  F = []
  for k in range(N): #Evaluamos fob para cada particula
    F.append(fob(Particulas[k,:]))
    if (F[k]>Fpbest[k]): #Actualizamos el mejor local
      Fpbest[k] = F[k]
      pbest[k,:] = Particulas[k,:].copy()

  indBest = np.argmax(F)
  if (F[indBest]>Fgbest): #Actualizamos el mejor global
    Fgbest = F[indBest]
    gbest = Particulas[indBest,:]

print('x: ',gbest,', f(x): ',fob(gbest))

x:  [-0.65045902 -0.34889855] , f(x):  -0.5448271380974716


In [42]:
A = 5*np.random.rand(10,2)-2

In [43]:
Ap = np.clip(A,[-2,-1],[1,2])

In [44]:
A,Ap

(array([[-0.80549786,  2.4472178 ],
        [ 1.85517382, -0.92556713],
        [ 0.69166613, -0.52943257],
        [-1.40035615, -1.92909613],
        [ 2.98835986, -0.22582036],
        [ 2.89610935, -1.31627657],
        [-1.76206774,  2.71515767],
        [-1.9351669 ,  2.82282034],
        [ 1.13706013, -0.68901115],
        [ 1.93927553,  2.40531953]]),
 array([[-0.80549786,  2.        ],
        [ 1.        , -0.92556713],
        [ 0.69166613, -0.52943257],
        [-1.40035615, -1.        ],
        [ 1.        , -0.22582036],
        [ 1.        , -1.        ],
        [-1.76206774,  2.        ],
        [-1.9351669 ,  2.        ],
        [ 1.        , -0.68901115],
        [ 1.        ,  2.        ]]))